src/erlcloud_config.erl

```diff
@@ -0,0 +1,632 @@
+new(AccessKeyID, SecretAccessKey, Host) ->
+    #aws_config{access_key_id     = AccessKeyID,
+                secret_access_key = SecretAccessKey,
+                config_host       = Host,
+                retry             = fun erlcloud_retry:default_retry/1}.
+
+-spec new(string(), string(), string(), non_neg_integer()) -> aws_config().
+new(AccessKeyID, SecretAccessKey, Host, Port) ->
+    #aws_config{access_key_id     = AccessKeyID,
+                secret_access_key = SecretAccessKey,
+                config_host       = Host,
+                config_port       = Port,
+                retry             = fun erlcloud_retry:default_retry/1}.
+
+-spec new(string(), string(), string(), non_neg_integer(), string()) ->
+    aws_config().
+new(AccessKeyID, SecretAccessKey, Host, Port, Scheme) ->
+    #aws_config{access_key_id     = AccessKeyID,
+                secret_access_key = SecretAccessKey,
+                config_host       = Host,
+                config_port       = Port,
+                config_scheme     = Scheme,
+                retry             = fun erlcloud_retry:default_retry/1}.
+
+-spec configure(string(), string()) -> ok.
+configure(AccessKeyID, SecretAccessKey) ->
+    put(aws_config, new(AccessKeyID, SecretAccessKey)),
+    ok.
+
+-spec configure(string(), string(), string()) -> ok.
+configure(AccessKeyID, SecretAccessKey, Host) ->
+    put(aws_config, new(AccessKeyID, SecretAccessKey, Host)),
+    ok.
+
+-spec configure(string(), string(), string(), non_neg_integer()) -> ok.
+configure(AccessKeyID, SecretAccessKey, Host, Port) ->
+    put(aws_config, new(AccessKeyID, SecretAccessKey, Host, Port)),
+    ok.
+
+-spec configure(string(), string(), string(), non_neg_integer(), string()) -> ok.
+configure(AccessKeyID, SecretAccessKey, Host, Port, Scheme) ->
+    put(aws_config, new(AccessKeyID, SecretAccessKey, Host, Port, Scheme)),
+    ok.
+
+%%------------------------------------------------------------------------------
+%% AWS Config API Functions
+%%------------------------------------------------------------------------------
+
+%%------------------------------------------------------------------------------
+%% @doc
+%% DescribeComplianceByConfigRule
+%%
+%% https://docs.aws.amazon.com/config/latest/APIReference/API_DescribeComplianceByConfigRule.html
+%%------------------------------------------------------------------------------
+-spec describe_compliance_by_config_rule() ->
+    result_paged(map()).
+describe_compliance_by_config_rule() ->
+    describe_compliance_by_config_rule(#{}).
+
+-spec describe_compliance_by_config_rule(Opts :: map()) ->
+    result_paged(map()).
+describe_compliance_by_config_rule(Opts) ->
+    describe_compliance_by_config_rule(Opts, default_config()).
+
+-spec describe_compliance_by_config_rule(Opts :: map(),
+    Config :: aws_config()) ->
+    result_paged(map()).
+describe_compliance_by_config_rule(Opts, Config) when is_map(Opts) orelse Opts =:= #{} ->
```

Comment:
I would say it is not a good idea in general to have a map (`Opts` in this case) as an input parameter.  In order to use this method, I need look into the spec and then search below in maps. 
```
+    Request = #{?COMPLIANCE_TYPES => maps:get(?COMPLIANCE_TYPES, Opts, []),
+        ?CONFIG_RULE_NAMES        => maps:get(?CONFIG_RULE_NAMES, Opts, []),
+        ?NEXT_TOKEN_LABEL         => maps:get(?NEXT_TOKEN_LABEL, Opts, <<"">>)
+    }
```

There are only macros thus I have to go through macros to find how to call parameters in my input map. Moreover, I can find what is in a map however it doesn't contain data type of values. 

As an user of this interface, I see it a bit difficult. I think it would be better to have all method parameters instead of a single map.

Reply:
I infer it as a best practice based on extending the logic here to its logical conclusion: https://github.com/ikemsley/erlcloud/blame/master/README.md#L184

Personally I much prefer it this way otherwise functions get an explosion of brittle and unwieldy arguments that are not safe since its easy to mix up their order, and you still have to look up their meaning anyway!  At least this way you can rtfm on the Amazon doc.

Reply:
@ikemsley I don't know what is the best practice in erlcloud. I would say @motobob meant that `NextToken` and `NextMarker` should be in `Opts` not all input parameters in general. 
If it is recommended way then it is fine, I am just saying it is a bit difficult to follow.  

FYI @motobob

Reply:
It is not written in Best Practices doc (and I want to add it there;) that we should:

* use Requires parameters as function param
* use optional parameters (including pagination stuff) as Options (map or proplist)

This way it makes it clear what is optional and what is not, and you get checking of required parameters presence in the compile time! (You still have to check the order yourself though)

Reply:
Okay, I'll do it like that.  Most of the parameters are optional anyway so I'm not sure anything will change much.  However, I'll go through carefully and check.